source/common/ssl/connection_impl.cc

```diff
@@ -139,16 +135,25 @@ Network::ConnectionImpl::PostIoAction ConnectionImpl::doHandshake() {
 
 void ConnectionImpl::drainErrorQueue() {
   bool saw_error = false;
+  bool saw_counted_error = false;
   while (uint64_t err = ERR_get_error()) {
     if (!saw_error) {
-      ctx_.stats().connection_error_.inc();
       saw_error = true;
     }
+    if (ERR_GET_REASON(err) == SSL_R_PEER_DID_NOT_RETURN_A_CERTIFICATE) {
+      ctx_.stats().fail_verify_no_cert_.inc();
+      saw_counted_error = true;
+    } else if (ERR_GET_REASON(err) == SSL_R_CERTIFICATE_VERIFY_FAILED) {
+      saw_counted_error = true;
+    }
 
     ENVOY_CONN_LOG(debug, "SSL error: {}:{}:{}:{}", *this, err, ERR_lib_error_string(err),
                    ERR_func_error_string(err), ERR_reason_error_string(err));
     UNREFERENCED_PARAMETER(err);
   }
+  if (saw_error && !saw_counted_error) {
```

Comment:
This is intentional change to make the behavior more consistent (per commit message).

Previously, failed certificate verification would count towards `ssl.connection_error`, but wrong certificate hash and/or SAN mismatch would count only towards `ssl.fail_verify_*`.

With this change, all certificate verification failures count towards `ssl.fail_verify_*` and `ssl.connection_error` is used only for connection errors and handshake failures for reasons other than failed certificate verification.